handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilterRuleComparator.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ipfilter;
+
+import java.net.InetSocketAddress;
+import java.util.Comparator;
+
+final class IpSubnetFilterRuleComparator implements Comparator<Object> {
+
+    @Override
+    public int compare(Object o1, Object o2) {
+        if (o1 instanceof IpSubnetFilterRule) {
+            if (o2 instanceof IpSubnetFilterRule) {
+                return ((IpSubnetFilterRule) o1).compareTo((IpSubnetFilterRule) o2);
+            } else if (o2 instanceof InetSocketAddress) {
+                return ((IpSubnetFilterRule) o1).matches((InetSocketAddress) o2) ? 0 : -1;
+            }
+        }
+        return -1;
```

Comment:
This is a rather strange comparator, that skimps on the reciprocity rule, and only kind of makes sense in this narrow use case. And then it also relies on the implementation details of `Collections.binarySearch` a bit. It's also not clear if it actually works as intended. I'd rather see a bespoke and well tested implementation of binary search as a helper method for `IpSubnetFilter.accept`, that incorporates the special comparison logic.

Reply:
Binary search requires me to pass same class type as key. So if I wanna search InetSocketAddress in list of IpSubnetFilterRule, it's not possible. So I used Object here.

o1 will always be IpSubnetFilterRule.
o2 will be:

1. IpSubnetFilterRule if we're sorting. (It'll use networkAddress for sort)

2. InetSocketAddress if we're searching. (It'll simple use "matches' method). 

You have a better approach for this?

Reply:
Yes, this particular use of binary search is a little unusual. That's why I suggested to either reimplement a bespoke binary search, that captures and handles the special comparison logic. Or, since the `Collections.sort` method has an overload that relies on `Comparable` rather than `Comparator`, we move or inline this `Comparator` to the `accept` (the only place it's used at that point) and simplify it (since o1 is now always a filter, and o2 is always an address).

Reply:
Also, I'd like to see a unit test that verifies that the `accept` method filters correctly; so everything works as intended.